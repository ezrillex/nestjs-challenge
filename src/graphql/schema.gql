# ------------------------------------------------------
# THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
# ------------------------------------------------------

type Categories {
  id: String!
  name: String!
}

type GetCategoriesResponse {
  id: String!
  name: String!
  products: [Products!]!
}

type Images {
  id: String!
  url: String!
  created_at: DateTime!
  product_variation_id: String
}

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type ProductVariations {
  id: String!
  title: String!
  price: Float!
  stock: Int!
  images: [Images!]
}

type Products {
  id: ID!
  is_published: Boolean!
  is_deleted: Boolean!
  name: String!
  description: String!
  variations: [ProductVariations!]!
  categories: [Categories!]!
  created_by: String!
  created_at: DateTime!
  last_updated_by: String
  last_updated_at: DateTime
}

type LikesOfProducts {
  user_id: ID!
  product_variation_id: ID!
  state: Boolean!
}

type Users {
  id: ID!
  first_name: String!
  last_name: String!
  email: String!
  role: String!
  created_at: DateTime!
  login_at: DateTime!
  password_last_updated: DateTime!
  likes_products: [LikesOfProducts!]!
}

type CartItems {
  id: ID!
  cart_owner: Users!
  product_variation: ProductVariations!
  quantity: Int!
}

type OrderItems {
  id: ID!
  product_variation: ProductVariations
  quantity: Int!
  price_purchased_at: Float!
}

type PaymentIntents {
  id: ID!
  status: String
  created_at: DateTime!
  stripe_event_id: ID!
}

type Orders {
  id: ID!
  user: Users!
  payments: [PaymentIntents!]!
  paymentStatus: String!
  orderStatus: String!
  order_items: [OrderItems!]!
}

type Query {
  user(id: ID!): Users!
  getLikes: [ProductVariations!]
  getProducts(GetProductsInput: GetProductsInput!): [Products!]
  getProductById(id: String!): Products
  getCategories: [GetCategoriesResponse!]
  getCartItems: [CartItems!]
  getOrders(GetOrdersInput: GetOrdersInput!): [Orders!]
  getOrder(order_id: String!): Orders
}

input GetProductsInput {
  first: Int
  offset: Int
  categoryFilter: [ID!]
  search: String
  likedOnly: Boolean
  omitImages: Boolean
  sort: String
}

input GetOrdersInput {
  first: Int
  offset: Int
  client_id: ID
}

type Mutation {
  toggleLike(product_variation_id: String!): LikesOfProducts
  createProduct(CreateProductInput: CreateProductInput!): Products
  updateProduct(UpdateProductInput: UpdateProductInput!): Products
  deleteProduct(product_id: String!): String
  updateProductVariation(UpdateProductVariationInput: UpdateProductVariationInput!): ProductVariations
  createProductVariation(CreateProductVariationInput: CreateProductVariationInput!): ProductVariations
  deleteProductVariation(variation_id: String!): String
  createCategory(name: String!): Categories
  deleteCategory(id: ID!): String
  addToCart(variation_id: ID!, quantity: Int!): CartItems
  removeFromCart(product_variation_id: String!): String
  createOrder: Orders
}

input CreateProductInput {
  name: String!
  description: String!
  categories: [ID!]
  variations: [CreateVariationInput!]!
}

input CreateVariationInput {
  title: String!
  price: Float!
  stock: Float!
}

input UpdateProductInput {
  id: ID!
  name: String
  description: String
  categories: [ID!]
  is_published: Boolean
}

input UpdateProductVariationInput {
  id: ID!
  title: String
  price: Float
  stock: Int
}

input CreateProductVariationInput {
  product_id: ID!
  title: String!
  price: Float!
  stock: Int!
}