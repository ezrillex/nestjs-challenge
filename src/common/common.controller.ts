import { Controller, Get, HttpCode, HttpStatus } from '@nestjs/common';
import { Public } from './decorators/public.decorator';

@Controller('common')
export class CommonController {
  @HttpCode(HttpStatus.NO_CONTENT)
  @Public()
  @Get()
  async healthService(): Promise<void> {}
}
