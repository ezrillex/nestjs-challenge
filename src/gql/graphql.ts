
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */

export class GetProductsInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    categoryFilter?: Nullable<string[]>;
    search?: Nullable<string>;
    likedOnly?: Nullable<boolean>;
    omitImages?: Nullable<boolean>;
    sort?: Nullable<string>;
}

export class CreateProductInput {
    name: string;
    description: string;
    categories?: Nullable<string[]>;
    variations: CreateVariationInput[];
}

export class CreateVariationInput {
    title: string;
    price: number;
    stock: number;
}

export class UpdateProductInput {
    id: string;
    name?: Nullable<string>;
    description?: Nullable<string>;
    categories?: Nullable<string[]>;
    is_published?: Nullable<boolean>;
}

export class UpdateProductVariationInput {
    id: string;
    title?: Nullable<string>;
    price?: Nullable<number>;
    stock?: Nullable<number>;
}

export class CreateProductVariationInput {
    product_id: string;
    title: string;
    price: number;
    stock: number;
}

export class CreateCategoryInput {
    name: string;
}

export class Images {
    __typename?: 'Images';
    id: string;
    url: string;
    created_at: string;
    product_variation_id?: Nullable<string>;
}

export class ProductVariations {
    __typename?: 'ProductVariations';
    id: string;
    title: string;
    price: number;
    stock: number;
    images?: Nullable<Images[]>;
}

export class Categories {
    __typename?: 'Categories';
    id: string;
    name: string;
    products: Products[];
}

export class Products {
    __typename?: 'Products';
    id: string;
    is_published: boolean;
    is_deleted: boolean;
    name: string;
    description: string;
    likes: ProductVariations[];
    variations: ProductVariations[];
    categories: Categories[];
    created_by: string;
    created_at: DateTime;
    last_updated_by?: Nullable<string>;
    last_updated_at?: Nullable<string>;
}

export class Users {
    __typename?: 'Users';
    id: string;
    first_name: string;
    last_name: string;
    email: string;
    likes_products: LikesOfProducts[];
}

export class LikesOfProducts {
    __typename?: 'LikesOfProducts';
    id: string;
    liked_by: Users;
    likes_product_variation: ProductVariations;
}

export class CartItems {
    __typename?: 'CartItems';
    id: string;
    cart_owner: Users;
    product_variation: ProductVariations;
    quantity: number;
}

export class OrderItems {
    __typename?: 'OrderItems';
    id: string;
    product_variation: ProductVariations;
    quantity: number;
    price_purchased_at: number;
}

export class Orders {
    __typename?: 'Orders';
    id: string;
    user: Users;
    paymentStatus: string;
    orderStatus: string;
    order_items: OrderItems[];
}

export abstract class IQuery {
    __typename?: 'IQuery';

    abstract getProducts(GetProductsInput: GetProductsInput): Nullable<Products[]> | Promise<Nullable<Products[]>>;

    abstract getProductById(id: string): Nullable<Products> | Promise<Nullable<Products>>;

    abstract getLikes(): Nullable<LikesOfProducts[]> | Promise<Nullable<LikesOfProducts[]>>;

    abstract getCartItems(): Nullable<CartItems[]> | Promise<Nullable<CartItems[]>>;

    abstract getOrders(): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;

    abstract getOrder(order_id: string): Nullable<Orders> | Promise<Nullable<Orders>>;

    abstract getClientOrders(client_id?: Nullable<string>): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;
}

export abstract class IMutation {
    __typename?: 'IMutation';

    abstract createProduct(CreateProductInput: CreateProductInput): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProduct(UpdateProductInput: UpdateProductInput): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProductVariation(UpdateProductVariationInput: UpdateProductVariationInput): Nullable<string> | Promise<Nullable<string>>;

    abstract createProductVariation(CreateProductVariationInput: CreateProductVariationInput): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteProductVariation(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteProduct(product_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createCategory(CreateCategoryInput: CreateCategoryInput): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteCategory(id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract likeProduct(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract removeLikeProduct(like_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract addToCart(variation_id: string, quantity: number): Nullable<string> | Promise<Nullable<string>>;

    abstract removeFromCart(cart_item_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createOrder(): Nullable<string> | Promise<Nullable<string>>;
}

export type DateTime = any;
type Nullable<T> = T | null;
