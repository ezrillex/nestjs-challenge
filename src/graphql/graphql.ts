
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */

export class GetProductsInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    categoryFilter?: Nullable<string[]>;
    search?: Nullable<string>;
    likedOnly?: Nullable<boolean>;
    omitImages?: Nullable<boolean>;
    sort?: Nullable<string>;
}

export class GetOrdersInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    client_id?: Nullable<string>;
}

export class CreateProductInput {
    name: string;
    description: string;
    categories?: Nullable<string[]>;
    variations: CreateVariationInput[];
}

export class CreateVariationInput {
    title: string;
    price: number;
    stock: number;
}

export class UpdateProductInput {
    id: string;
    name?: Nullable<string>;
    description?: Nullable<string>;
    categories?: Nullable<string[]>;
    is_published?: Nullable<boolean>;
}

export class UpdateProductVariationInput {
    id: string;
    title?: Nullable<string>;
    price?: Nullable<number>;
    stock?: Nullable<number>;
}

export class CreateProductVariationInput {
    product_id: string;
    title: string;
    price: number;
    stock: number;
}

export class Categories {
    __typename?: 'Categories';
    id: string;
    name: string;
}

export class GetCategoriesResponse {
    __typename?: 'GetCategoriesResponse';
    id: string;
    name: string;
    products: Products[];
}

export class Images {
    __typename?: 'Images';
    id: string;
    url: string;
    created_at: DateTime;
    product_variation_id?: Nullable<string>;
}

export class ProductVariations {
    __typename?: 'ProductVariations';
    id: string;
    title: string;
    price: number;
    stock: number;
    images?: Nullable<Images[]>;
}

export class Products {
    __typename?: 'Products';
    id: string;
    is_published: boolean;
    is_deleted: boolean;
    name: string;
    description: string;
    variations: ProductVariations[];
    categories: Categories[];
    created_by: string;
    created_at: DateTime;
    last_updated_by?: Nullable<string>;
    last_updated_at?: Nullable<DateTime>;
}

export class LikesOfProducts {
    __typename?: 'LikesOfProducts';
    user_id: string;
    product_variation_id: string;
    state: boolean;
}

export class Users {
    __typename?: 'Users';
    id: string;
    first_name: string;
    last_name: string;
    email: string;
    role: string;
    created_at: DateTime;
    login_at: DateTime;
    password_last_updated: DateTime;
    likes_products: LikesOfProducts[];
}

export class CartItems {
    __typename?: 'CartItems';
    id: string;
    cart_owner: Users;
    product_variation: ProductVariations;
    quantity: number;
}

export class OrderItems {
    __typename?: 'OrderItems';
    id: string;
    product_variation: ProductVariations;
    quantity: number;
    price_purchased_at: number;
}

export class PaymentIntents {
    __typename?: 'PaymentIntents';
    id: string;
    status?: Nullable<string>;
    created_at: DateTime;
    stripe_event_id: string;
}

export class Orders {
    __typename?: 'Orders';
    id: string;
    user: Users;
    payments: PaymentIntents[];
    paymentStatus: string;
    orderStatus: string;
    order_items: OrderItems[];
}

export abstract class IQuery {
    __typename?: 'IQuery';

    abstract user(id: string): Users | Promise<Users>;

    abstract getLikes(): Nullable<ProductVariations[]> | Promise<Nullable<ProductVariations[]>>;

    abstract getProducts(GetProductsInput: GetProductsInput): Nullable<Products[]> | Promise<Nullable<Products[]>>;

    abstract getProductById(id: string): Nullable<Products> | Promise<Nullable<Products>>;

    abstract getCategories(): Nullable<GetCategoriesResponse[]> | Promise<Nullable<GetCategoriesResponse[]>>;

    abstract getCartItems(): Nullable<CartItems[]> | Promise<Nullable<CartItems[]>>;

    abstract getOrders(GetOrdersInput: GetOrdersInput): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;

    abstract getOrder(order_id: string): Nullable<Orders> | Promise<Nullable<Orders>>;
}

export abstract class IMutation {
    __typename?: 'IMutation';

    abstract toggleLike(product_variation_id: string): Nullable<LikesOfProducts> | Promise<Nullable<LikesOfProducts>>;

    abstract createProduct(CreateProductInput: CreateProductInput): Nullable<Products> | Promise<Nullable<Products>>;

    abstract updateProduct(UpdateProductInput: UpdateProductInput): Nullable<Products> | Promise<Nullable<Products>>;

    abstract deleteProduct(product_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProductVariation(UpdateProductVariationInput: UpdateProductVariationInput): Nullable<ProductVariations> | Promise<Nullable<ProductVariations>>;

    abstract createProductVariation(CreateProductVariationInput: CreateProductVariationInput): Nullable<ProductVariations> | Promise<Nullable<ProductVariations>>;

    abstract deleteProductVariation(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createCategory(name: string): Nullable<Categories> | Promise<Nullable<Categories>>;

    abstract deleteCategory(id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract addToCart(variation_id: string, quantity: number): Nullable<CartItems> | Promise<Nullable<CartItems>>;

    abstract removeFromCart(product_variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createOrder(): Nullable<Orders> | Promise<Nullable<Orders>>;
}

export type DateTime = any;
type Nullable<T> = T | null;
