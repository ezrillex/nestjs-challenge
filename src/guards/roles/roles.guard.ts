import {
  CanActivate,
  ExecutionContext,
  ForbiddenException,
  Injectable,
} from '@nestjs/common';
import { Observable } from 'rxjs';
import { Reflector } from '@nestjs/core';
import { ROLE_REQUIRED } from '../../decorators/requires-role/requires-role.decorator';

@Injectable()
export class RolesGuard implements CanActivate {
  constructor(private reflector: Reflector) {}

  canActivate(
    context: ExecutionContext,
  ): boolean | Promise<boolean> | Observable<boolean> {
    console.log('Roles Guard triggered.');
    const rolesRequired = this.reflector.getAllAndOverride(ROLE_REQUIRED, [
      context.getHandler(),
      context.getClass(),
    ]);
    if (!rolesRequired) {
      return true;
    }

    const request = context.switchToHttp().getRequest();

    if (request.user && request.user.role === rolesRequired) {
      return true;
    } else {
      throw new ForbiddenException(
        'User lacks the necessary role for performing the action.',
      );
    }
  }
}
