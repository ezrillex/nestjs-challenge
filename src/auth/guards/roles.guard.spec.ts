import { Test, TestingModule } from '@nestjs/testing';
import { RolesGuard } from './roles.guard';
import { Reflector } from '@nestjs/core';
import { ExecutionContext } from '@nestjs/common';
import { roles } from '@prisma/client';
import { GqlExecutionContext } from '@nestjs/graphql';

describe('Roles Guard', () => {
  let guard: RolesGuard;
  let reflector: Reflector;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [RolesGuard, Reflector],
    }).compile();

    guard = module.get<RolesGuard>(RolesGuard);
    reflector = module.get<Reflector>(Reflector);
  });

  it('should be defined', () => {
    expect(guard).toBeDefined();
  });

  it('should implement can activate', () => {
    expect(guard.canActivate).toBeInstanceOf(Function);
  });

  it('should be true if it cant get any metadata set by decorator that specifies roles that are required', () => {
    jest.spyOn(reflector, 'getAllAndOverride').mockReturnValue(null);
    expect(
      guard.canActivate({
        getHandler: jest.fn(),
        getClass: jest.fn(),
      } as unknown as ExecutionContext),
    ).toEqual(true);
  });

  it('throws if user lacks the role required', () => {
    jest.spyOn(reflector, 'getAllAndOverride').mockReturnValue(roles.manager);
    expect(() => {
      guard.canActivate({
        getHandler: jest.fn(),
        getClass: jest.fn(),
        getType: jest.fn().mockReturnValue('http'),
        switchToHttp: jest.fn().mockReturnValue({
          getRequest: jest.fn().mockReturnValue({
            user: {
              role: roles.customer,
            },
          }),
        }),
      } as unknown as ExecutionContext);
    }).toThrowErrorMatchingSnapshot('insufficient permissions');
  });

  it('is true if sufficient permissions', () => {
    jest.spyOn(reflector, 'getAllAndOverride').mockReturnValue(roles.customer);
    expect(
      guard.canActivate({
        getHandler: jest.fn(),
        getClass: jest.fn(),
        getType: jest.fn().mockReturnValue('http'),
        switchToHttp: jest.fn().mockReturnValue({
          getRequest: jest.fn().mockReturnValue({
            user: {
              role: roles.customer,
            },
          }),
        }),
      } as unknown as ExecutionContext),
    ).toEqual(true);
  });

  it('it is able to fetch data for graphql query, returns true if ok permissions', () => {
    jest.spyOn(reflector, 'getAllAndOverride').mockReturnValue(roles.customer);
    jest.spyOn(GqlExecutionContext, 'create').mockReturnValue({
      getContext: jest.fn().mockReturnValue({
        req: {
          user: {
            role: roles.customer,
          },
        },
      }),
    } as unknown as GqlExecutionContext);
    expect(
      guard.canActivate({
        getHandler: jest.fn(),
        getClass: jest.fn(),
        getType: jest.fn().mockReturnValue('graphql'),
      } as unknown as ExecutionContext),
    ).toEqual(true);
  });

  it('it is able to fetch data for graphql query, throws when insufficient permissions', () => {
    jest.spyOn(reflector, 'getAllAndOverride').mockReturnValue(roles.manager);
    jest.spyOn(GqlExecutionContext, 'create').mockReturnValue({
      getContext: jest.fn().mockReturnValue({
        req: {
          user: {
            role: roles.customer,
          },
        },
      }),
    } as unknown as GqlExecutionContext);
    expect(() => {
      guard.canActivate({
        getHandler: jest.fn(),
        getClass: jest.fn(),
        getType: jest.fn().mockReturnValue('graphql'),
      } as unknown as ExecutionContext);
    }).toThrowErrorMatchingSnapshot('insufficient permissions graphql');
  });
});
