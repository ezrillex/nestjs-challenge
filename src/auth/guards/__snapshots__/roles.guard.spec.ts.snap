// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roles Guard it is able to fetch data for graphql query, throws when insufficient permissions: insufficient permissions graphql 1`] = `"User lacks the necessary role for performing the action."`;

exports[`Roles Guard throws if user lacks the role required: insufficient permissions 1`] = `"User lacks the necessary role for performing the action."`;
